control-plane/agents/src/bin/core/controller/reconciler/nexus/mod.rs

```diff
@@ -130,61 +135,139 @@ async fn nexus_reconciler(
     }
 }
 
-/// Find and removes faulted children from the given nexus
-/// If the child is a replica it also disowns and destroys it
+/// Checks if nexus is Degraded and any child is Faulted. If yes, Depending on rebuild policy for
+/// child it performs rebuild operation. We exclude NoSpace Degrade.
 #[tracing::instrument(skip(nexus, context), level = "trace", fields(nexus.uuid = %nexus.uuid(), request.reconcile = true))]
-pub(super) async fn faulted_children_remover(
+pub(super) async fn handle_faulted_child(
     nexus: &mut OperationGuardArc<NexusSpec>,
     context: &PollContext,
 ) -> PollResult {
     let nexus_uuid = nexus.uuid();
     let nexus_state = context.registry().nexus(nexus_uuid).await?;
     let child_count = nexus_state.children.len();
-
-    // Remove faulted children only from a degraded nexus with other healthy children left
     if nexus_state.status == NexusStatus::Degraded && child_count > 1 {
-        let span = tracing::info_span!("faulted_children_remover", nexus.uuid = %nexus_uuid, request.reconcile = true);
-        async {
-            let nexus_spec_clone = nexus.lock().clone();
-            for child in nexus_state.children.iter().filter(|c| c.state.faulted()) {
-                nexus_spec_clone.warn_span(|| {
-                    tracing::warn!(%child.state, %child.state_reason, "Attempting to remove faulted child '{}'", child.uri)
-                });
-                if let Err(error) = nexus
-                    .remove_child_by_uri(
-                        context.registry(),
-                        &nexus_state,
-                        &child.uri,
-                        true,
-                    )
+        let nexus_spec_clone = nexus.lock().clone();
+        for child in nexus_state.children.iter().filter(|c| c.state.faulted()) {
+            if child.state_reason != ChildStateReason::NoSpace {
+                let wait_time = RuleSet::faulted_child_wait(&nexus_state, context.registry());
```

Comment:
Please combine first two branches - 

 ```
             let elapsed = is_time_elapsed(child.faulted_at, Some(wait_time), &child.uri).await;
              if wait_time.is_zero()
                    || child.state_reason.clone() == ChildStateReason::RebuildFailed
                    || elapsed
                {
                    info!(%child.uri, "Start full rebuild for child, wait elapsed {elapsed}");
                    faulted_children_remover(nexus, &child.uri, context).await?
                } 
```

Reply:
It can be done but we will lose the logger in that case when is_time_elapsed() is true. Without a separate condition. We do have a log in is_time_elapsed() that lets us know how much we waited and wait_duration.

Reply:
We won't lose logger, I've taken care of that in suggestion.
`info!(%child.uri, "Start full rebuild for child, wait elapsed {elapsed}");`

Reply:
This log doesnt hold true for RebuildFailed child right? We do get duration from policy for that child but dont really care about time_elapsed for that child. elapsed can very well be false in this scenario.

Reply:
info!(%child.uri, "Start full rebuild for child, wait elapsed {elapsed}, child state reason: {}", child.state_reason);

I think this would help us figure out why we started Full rebuild